packages/arb-avm-cpp/data_storage/src/arbcore.cpp

```diff
@@ -896,6 +970,14 @@ ArbCore::loadLastMatchingMachine(
     ReadWriteTransaction& tx,
     std::unique_ptr<rocksdb::Iterator>& checkpoint_it,
     ValueCache& value_cache) {
+    // First check if last saved machine matches
+    auto mach = combined_machine_cache.checkSimpleMatching(check_output);
+    if (mach.machine != nullptr) {
+        // Found machine in cache
+        return std::make_unique<MachineThread>(mach.machine->machine_state);
```

Comment:
Do we have a guarantee that this is past the last checkpoint?

Reply:
Every time machine stops, one of two things happens:
1. machine added to basic cache, and last*_cache is  cleared if new machine is newer
2. last_cache is updated with new machine

Only after cache is updated will the decision to create new checkpoints is done.

If reorg happens, out of date last*_cache are cleared.

Given all of that, checkSimpleMatching is guaranteed to always b at or past the last checkpoint, or empty.

Reply:
Added extra check to only use last cache if it is more up to date than database